### PR TITLE
MLeap Flavor/Deployment #2: Python support for the MLeap flavor

### PR DESCRIPTION
**Pull request by dbczumar**
This is the **second** PR in a multi-part series that will introduce support for the MLeap flavor in MLFlow. 

This PR introduces a Python module for serializing PySpark MLLib models in MLeap format, `mleap.py`. It also modifies the existing `spark.py` module to produce models with the MLeap flavor when MLeap-compatible serialization parameters are specified. Finally, this also unifies `exceptions.py` and `utils/exception.py`.

For context about how these changes fit into the overall MLeap deployment integration, see #278

### CODE REVIEW

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good, thanks Corey!

I left some (minor) comments.

File: tests/spark/test_spark_model_export.py [hunk 1]
```diff
@@ -57,102 +59,214 @@ def test_hadoop_filesystem(tmpdir):
...
                 shutil.rmtree(x)
             finally:
-                mlflow.end_run()
-                mlflow.set_tracking_uri(old_tracking_uri)
+                tracking.end_run()
+                tracking.set_tracking_uri(old_tracking_uri)
                 shutil.rmtree(tracking_dir)
+
+
+def test_spark_module_model_save_without_sample_output_produces_sparkml_flavor(
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply tests/spark/test_spark_model_export.py]: Removed redundant test and replaced it with a test for `spark.log_model`

File: tests/spark/test_spark_model_export.py [hunk 2]
```diff
@@ -57,102 +59,214 @@ def test_hadoop_filesystem(tmpdir):
...
+                      mlflow_model=mlflow_model)
+    assert sparkm.FLAVOR_NAME in mlflow_model.flavors
+
+    config_path = os.path.join(model_path, "MLmodel")
+    assert os.path.exists(config_path)
+    config = Model.load(config_path)
+    assert sparkm.FLAVOR_NAME in config.flavors
+
+
+def test_spark_module_model_save_with_sample_output_produces_sparkml_and_mleap_flavors(
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply tests/spark/test_spark_model_export.py]: Good idea. Added the following test cases: `test_mleap_module_model_save_with_unsupported_transformer_raises_exception`, `test_spark_module_model_save_with_mleap_and_unsupported_transformer_raises_exception`

File: mlflow/mleap.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                  cannot contain any custom transformers.
+    :param sample_input: A sample PySpark Dataframe input that the model can evaluate. This is 
+                         required by MLeap for data schema inference.
+    :param mlflow_model: MLFlow model config to which this flavor is being added.
+    """
+    _add_to_model(mlflow_model, path, spark_model, sample_input)
+    mlflow_model.save(os.path.join(path, "MLmodel"))
+
+
+def _add_to_model(mlflow_model, path, spark_model, sample_input):
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/mleap.py]: Sure. I had initially made it private due to this difference in functionality, but I think it makes sense to publicly expose the method.

**Review by dbczumar — COMMENTED** [review]
@tomasatdatabricks Addressed your comments!

**Review by tomasatdatabricks — APPROVED** [review]
Looks great, thanks Corey!

**Pull request merged**